lib/internal/child_process.js

```diff
@@ -733,27 +738,32 @@ function setupChannel(target, channel) {
     process.nextTick(finish);
   };
 
+  function emit(event, message, handle) {
+    target.emit(event, message, handle);
+  }
+
+  function handleMessage(message, handle, internal, nextTick) {
+    if (!target.channel)
+      return;
+
+    var eventName = (internal ? 'internalMessage' : 'message');
+    if (nextTick)
+      process.nextTick(emit, eventName, message, handle);
+    else
+      target.emit(eventName, message, handle);
+  }
+
   channel.readStart();
   return control;
 }
 
-
 const INTERNAL_PREFIX = 'NODE_';
-function handleMessage(target, message, handle) {
-  if (!target.channel)
-    return;
-
-  var eventName = 'message';
-  if (message !== null &&
-      typeof message === 'object' &&
-      typeof message.cmd === 'string' &&
-      message.cmd.length > INTERNAL_PREFIX.length &&
-      message.cmd.slice(0, INTERNAL_PREFIX.length) === INTERNAL_PREFIX) {
-    eventName = 'internalMessage';
-  }
-  process.nextTick(() => {
-    target.emit(eventName, message, handle);
-  });
+function isInternal(message) {
+  return (message !== null &&
+          typeof message === 'object' &&
+          typeof message.cmd === 'string' &&
```

Comment:
[question] since internal messages are rare, will adding a `Object.hasOwnProperty(message, 'cmd') &&` fail faster than `typeof message.cmd === 'string'`?